### DISCUSSION TITLE
Create AKS Cluster Functionality 🥷💡 ❤️

### DISCUSSION THREAD

**Discussion by Tatsinnit** [Azure/vscode-aks-tools#204] [Ideas]
This idea is floated around few times in hallway chats and indirect conversations and pings, hence I am documenting what I've found for visibility and we can choose the direction we want to move forward.

I have preference towards doing it in 3 + ways discussed in Phases below.

* Phase 1 : Fix the existing kubernetes extension `Create Cluster`
    * Currently this behaviour is not working for me at least so sending a fix PR against this repo to begin with will be key.

* Phase 2 (Optional): Create AKS Cluster as Separate AKS Extension Behaviour. (AKS Exclusivity but kind of redoing what the above extension already does just differently 🤷🏽‍♂️
    * We can do this by building either with a brand new `webview` or `multi-step wizard` detail for both below.
    * We need this to be minimalistic so we can choose one of the delivery mechanism detail under following API f0r Create Cluster for AKS as drop-0/1. 
        * API detail: https://learn.microsoft.com/en-us/rest/api/aks/managed-clusters/create-or-update?tabs=HTTP
#### Implementation level for new wizard to choose from

* Details:
     * `Webviews` approach
          *  They are Accessible: https://code.visualstudio.com/api/extension-guides/webview#accessibility 
          * A very cryptic guideline here, so from this we can understand better with this excercise: https://code.visualstudio.com/api/ux-guidelines/webviews  
    * VsCode `multistep` wizard:
           *  Docs here: https://code.visualstudio.com/api/ux-guidelines/quick-picks 
           * Sample: https://code.visualstudio.com/api/ux-guidelines/quick-picks#multiple-steps  

* Hybrid: Less Confusing. 🥷💡
   * By this what I mean is to fix Kubernetes extension for creating AKS cluster scenario and might betook that command from the AKS Cluster navigation as well (using `when` clause. It will be simple and won't confuse users  

Thanks heaps in advance for taking time and reading this and humble fyi to cc: @squillace, @rzhang628, @peterbom, @gambtho 🥷💡 ❤️ 

**Screenshot 1***

<img width="649" alt="Screenshot 2023-02-22 at 8 18 01 AM" src="https://user-images.githubusercontent.com/6233295/220438829-b6895005-504e-4b97-a4d6-f9b4214af117.png">

**Comment by peterbom** [comment]
Thanks for pulling these ideas together, @Tatsinnit!

I wanted to throw in one more approach we have discussed previously. It may be possible to work with the portal team to improve the ability to deep-link into the 'create cluster' experience in the portal (at the moment I don't think it's possible to construct links that have any kind of pre-selection, even for subscription, but happy to be corrected if I'm wrong). At the very least we could pre-select subscription, but I wonder whether it'd be possible to create other pre-filled cluster 'profiles' of some kind, perhaps based on templates.

The obvious drawback of this is it would kick users out of the VSCode environment. But it'd be good to weigh that against the development and maintenance overhead of mirroring the portal's 'create cluster' experience in our extension, which I wouldn't want to underestimate.

Also, something I just learned from reading the docs above: VSCode's [multi-step quick-pick](https://code.visualstudio.com/api/ux-guidelines/quick-picks#multiple-steps) approach is not recommended for a large number of steps ("they aren't well suited to function as a wizard or similarly complex experience"). I think this counts against it as a candidate for the create-cluster workflow.

As a side-note, I'm pleased to hear that accessibility within webviews has been thought about in VSCode, although it'll take some effort on our part to build a good experience.

**Reply by Tatsinnit** [reply]
🥷 ❤️ 🐈 Absolutely yes - I am so glad you mentioned the Option-4 - `deep linking` portal - `Create Cluster` page - we do have something similar but it lands user to the `portal marketplace service - create cluster page`


<img width="423" alt="Screenshot 2023-02-22 at 4 37 42 PM" src="https://user-images.githubusercontent.com/6233295/220515360-eec71842-e4a9-4fb3-870f-cdf031d728ca.png">

**Comment by sabbour** [comment]
How about creating  "dev/test" optimized clusters (see preset configurations from portal)  where we create a minimal cluster to get started and then follow that by downloading the credentials and guiding them to create an app?

**Reply by Tatsinnit** [reply]
💡☕️❤️ Thank you for the comments and this sounds like a great idea! @sabbour I wonder if we can do something very minimalistic, here is sample json for attributes, and if you can help in filing what all fields we can default to what for `dev/test` aks cluster please. 🙏

* Given we have this, `resourceGroup, SubId, Get-> clusterName from user` what extra we need or below will suffice?
* Can we default `location` to a specific region?  (thinking out loud, otherwise - will that involve any combinations to serve some sort of list for the user to select from?)

Thanks heaps.

```*ts
const parameters = {
    addonProfiles: {},
    agentPoolProfiles: [
      {
        name: "nodepool1",
        type: "VirtualMachineScaleSets",
        count: 3,
        creationData: {
          sourceResourceId:
            "/subscriptions/subid1/resourceGroups/rg1/providers/Microsoft.ContainerService/snapshots/snapshot1",
        },
        enableFips: true,
        enableNodePublicIP: true,
        mode: "System",
        osType: "Linux",
        vmSize: "Standard_DS2_v2",
      },
    ],
    autoScalerProfile: { scaleDownDelayAfterAdd: "15m", scanInterval: "20s" },
    diskEncryptionSetID:
      "/subscriptions/subid1/resourceGroups/rg1/providers/Microsoft.Compute/diskEncryptionSets/des",
    dnsPrefix: "dnsprefix1",
    enablePodSecurityPolicy: false,
    enableRbac: true,
    kubernetesVersion: "",
    linuxProfile: {
      adminUsername: "azureuser",
      ssh: { publicKeys: [{ keyData: "keydata" }] },
    },
    location: "location1",
    networkProfile: {
      loadBalancerProfile: { managedOutboundIPs: { count: 2 } },
      loadBalancerSku: "standard",
      outboundType: "loadBalancer",
    },
    servicePrincipalProfile: { clientId: "clientid", secret: "secret" },
    sku: { name: "Basic", tier: "Free" },
    tags: { archv2: "", tier: "production" },
    windowsProfile: {
      adminPassword: "replacePassword1234$",
      adminUsername: "azureuser",
    },
  };
```

**Reply by Tatsinnit** [reply]
Actually seems like we could do this as first drop:

### create-managed-cluster-with-ossku

https://learn.microsoft.com/en-us/rest/api/aks/managed-clusters/create-or-update?tabs=HTTP#create-managed-cluster-with-ossku


```*.json
PUT https://management.azure.com/subscriptions/subid1/resourceGroups/rg1/providers/Microsoft.ContainerService/managedClusters/clustername1?api-version=2023-04-01

{
  "location": "location1",
  "tags": {
    "tier": "production",
    "archv2": ""
  },
  "sku": {
    "name": "Basic",
    "tier": "Free"
  },
  "properties": {
    "kubernetesVersion": "",
    "dnsPrefix": "dnsprefix1",
    "agentPoolProfiles": [
      {
        "name": "nodepool1",
        "count": 3,
        "vmSize": "Standard_DS2_v2",
        "osType": "Linux",
        "osSKU": "AzureLinux",
        "type": "VirtualMachineScaleSets",
        "enableNodePublicIP": true,
        "mode": "System"
      }
    ],
    "linuxProfile": {
      "adminUsername": "azureuser",
      "ssh": {
        "publicKeys": [
          {
            "keyData": "keydata"
          }
        ]
      }
    },
    "networkProfile": {
      "loadBalancerSku": "standard",
      "outboundType": "loadBalancer",
      "loadBalancerProfile": {
        "managedOutboundIPs": {
          "count": 2
        }
      }
    },
    "autoScalerProfile": {
      "scan-interval": "20s",
      "scale-down-delay-after-add": "15m"
    },
    "windowsProfile": {
      "adminUsername": "azureuser",
      "adminPassword": "replacePassword1234$"
    },
    "servicePrincipalProfile": {
      "clientId": "clientid",
      "secret": "secret"
    },
    "addonProfiles": {},
    "enableRBAC": true,
    "diskEncryptionSetID": "/subscriptions/subid1/resourceGroups/rg1/providers/Microsoft.Compute/diskEncryptionSets/des",
    "enablePodSecurityPolicy": true,
    "httpProxyConfig": {
      "httpProxy": "http://myproxy.server.com:8080",
      "httpsProxy": "https://myproxy.server.com:8080",
      "noProxy": [
        "localhost",
        "127.0.0.1"
      ],
      "trustedCa": "Q29uZ3JhdHMhIFlvdSBoYXZlIGZvdW5kIGEgaGlkZGVuIG1lc3NhZ2U="
    }
  }
}
```

**Reply by peterbom** [reply]
Seems to contain quite a few properties I'd expect to be defaulted or omitted? E.g:
- `tags`
- `dnsPrefix`
- `windowsProfile`
- `servicePrincipalProfile`
- `httpProxyConfig`

We probably wouldn't be wanting the user to enter any credentials or use the SSH keys on their machine, I guess?

**Comment by Tatsinnit** [comment]
Update: 

* A good drop of this work reside here: https://github.com/Azure/vscode-aks-tools/pull/217 

Thanks heaps.
